src/common/meta/src/cluster.rs

```diff
@@ -69,6 +71,28 @@ pub struct NodeInfoKey {
 }
 
 impl NodeInfoKey {
+    /// Try to create a `NodeInfoKey` from a "good" heartbeat request. "good" as in every needed
+    /// piece of information is provided and valid.  
+    pub fn new(request: &HeartbeatRequest) -> Option<Self> {
+        let HeartbeatRequest { header, peer, .. } = request;
+        let header = header.as_ref()?;
+        let peer = peer.as_ref()?;
+
+        let role = header.role.try_into().ok()?;
+        let node_id = match role {
+            // Because the Frontend is stateless, it's too easy to neglect choosing a unique id
+            // for it when setting up a cluster. So we calculate its id from its address.
+            Role::Frontend => calculate_node_id(&peer.addr),
```

Comment:
This relies on the `peer.addr` is correctly detected. 
@zyy17 wdyt

Reply:
I think we should set the frontend's id correctly at the first place, can be simply calculated out from its pod address at the operator side